ANARI

griffin28 opened the pull request:
### Description

Resolves #18416

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->

Add new surface rendering capability using [ANARI](https://github.com/KhronosGroup/ANARI-SDK). Note, this PR doesn't include the dynamic UI work since the code to support it is currently on VTK's master branch. Talked with Kitware and they said it will be in the **VTK 9.5.0** release targeted for the summer 2025. 

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* ~~[ ] Bug fix~~
* [X] New feature
* [X] Documentation update
* ~~[ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?
1. Open a test dataset in VisIt (e.g., jet.cgns)
2. Create a Pseudocolor plot of the variable `CoefPressure`
3. Open the Rendering options window and select **ANARI Rendering** and options as shown below.
4. Click apply and now you're rendering with ANARI!
5. Deselect **ANARI Rendering**, apply, and now you're back to VisIt's default rendering.

![helide_test](https://github.com/user-attachments/assets/b4a125c1-f9d2-4bc9-9ee2-c9c7c5a890a1)

### Reminders:

- Please follow the [style guidelines][1] of this project.
- Please perform a self-review of your code before submitting a PR and asking others to review it.
- Please assign reviewers (see [VisIt's PR procedures][2] for more information).

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have commented my code where applicable.
- [X] I have updated the release notes.
- [X] I have made corresponding changes to the documentation.
- ~~[ ] I have added debugging support to my changes.~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- [X] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

JustinPrivitera committed:
Fix infinite loop in `has_nonspace_chars` (#20294)

markcmiller86 committed:
Add GetActiveWindow (#20296) (#20298)

markcmiller86 committed:
Add more plotting options by topological dimension (#20285) (#20300)

griffin28 committed:
ANARI integration

cyrush committed:
fix issue excluding visit-install script from release checksums (#20301)

griffin28 committed:
Updated User Manual and release notes

biagas commented on src/avt/Pipeline/Data/avtCallback.C:
```diff
@@ -33,6 +33,9 @@ bool                          avtCallback::safeMode = false;
 #if defined(HAVE_OSPRAY)
 bool                          avtCallback::useOSPRay = false;
 #endif
+#ifdef VISIT_ANARI
```
Does `VISIT_ANARI` indicate user has chosen anari at runtime via `-anari`? Or just the existence of Anari?  If the latter, it seems `HAVE_ANARI` would be the better choice.

griffin28 replied:
The latter...changing to `HAVE_ANARI`

biagas commented on src/viewer/main/VisItViewer.C:
```diff
@@ -93,6 +93,13 @@ VisItViewer::Initialize(int *argc, char ***argv)
         }
+#endif
+#ifdef VISIT_ANARI
+        else if (strcmp((*argv)[i], "-anari") == 0)
```
Does adding `-anari` to the command line update the RenderingAttributes `Anari` checkbox?
Can it only be enabled on startup, or is choosing the option via the RenderingAttributes sufficient?
Just wondering why the need for the command line option?

griffin28 replied:
Not needed, I removed this. I was just copying OSPRay.

biagas committed:
Remove use of sprintf, as it was causing the crash. (#20306)

biagas committed:
Fix unecessary thirdparty tarball downloads. (#20313)

Fix calls to bv_ensure_built_or_ready.
Add ensure_built_or_ready_component to test for existence of
individual libs/components.
Modified bv_xcb to use ensure_built_or_ready_component.

biagas committed:
Enable osmesa diagnostic tests when building with VTK 9.4.1. (#20315)

brugger1 committed:
Fix the MOAB test failures. (#20320)

biagas committed:
Remove ultrawrapper. (#20308)

* Remove ultrawrapper.

* Remove pyparsing from list of TP downloads.

brugger1 committed:
Update the last test suite pass on poodle.

biagas committed:
For QLineEdit, switch connections from `textChanged` to `editingFinished`. (#20322)

Reason: textChanged is a signal emitted with every keystroke. In most instances we only want the signal when all text has been entered.
There are a few exceptions: QLineEdit with validator attached, or a `search` widget that updates displayed contents as user types. Those instances were not modified.

brugger1 committed:
Update the last test suite pass on poodle.

brugger1 committed:
Update the last test suite pass on poodle.

markcmiller86 committed:
Apply patch for STORE_SETS_FIELDS (#20333)

Fixed issues causing test to fail. Removed associated moab test from skip list.

markcmiller86 commented:
@griffin28 haven't forgotten. Have a few outstanding PRs to get to. Will likely get to by tomorrow PM.

griffin28 committed:
Updates based on PR feedback and some clean-up

griffin28 commented:
My apologies @markcmiller86  and @biagas I did a rebase on the latest changes from develop before I pushed my updates. I should've waited until the review was done because now it's showing all of those other commits (137 files changed!!).

griffin28 commented:
The [first commit](https://github.com/visit-dav/visit/pull/20303/commits/f901aa977d3fc33f9bb57ad49df1937c159b19de) will have all the files I've actually touched and the [last commit](https://github.com/visit-dav/visit/pull/20303/commits/8d2ffa9713398b374b08a5ab52011390100ffae1) will have the current updates based on @biagas feedback.

biagas commented:
> I did a rebase on the latest changes from develop

Sometimes a merge-from-develop into the branch works better than a rebase. No worries, it still workable.

markcmiller86 commented on src/avt/Plotter/CMakeLists.txt:
```diff
@@ -145,17 +148,18 @@ endif()
     visitcommon avtmath avtview avtfilters_ser avtpipeline_ser visit_vtk
-    lightweight_visit_vtk ${vtklibs} ${OPENGL_LIBRARIES})
+    lightweight_visit_vtk ${vtklibs} ${OPENGL_LIBRARIES}
+    $<$<BOOL:${VISIT_ANARI}>:VTK::RenderingAnari>)
```
Just a note to self...this is using CMake expression syntax. I see this in a couple of places in BLT too.

markcmiller86 commented on src/avt/Plotter/vtk/vtkAnariVisItAxisActorNode.C:
I notice we have one of these for Ospray too. What is driving the need for axis actors for these scenarios...translating camera info to the form needed in the renderer?

griffin28 replied:
For the VTK Factory overrides that VisIt is doing. This is to override VisIt's override so the axis rendering is going through ANARI when ANARI rendering is enabled.
```
factory->RegisterOverride("vtkVisItAxisActor", vtkAnariVisItViewNodeFactory::axis_act_maker);
```

markcmiller86 replied:
Ok. Makes sense.

markcmiller86 commented on src/avt/View/avtViewInfo.C:
```diff
@@ -316,12 +315,17 @@ avtViewInfo::SetCameraFromView(vtkCamera *vtkcam) const
     vtkcam->SetWindowCenter(2.0*imagePan[0], 2.0*imagePan[1]);
     vtkcam->SetFocalDisk(imageZoom);
 
+    #ifdef HAVE_ANARI
```
This block seems a tad funky...we're testing for anari at compile time but ospray at run time. Does that mean that if VisIt is compiled with ANARI, that ANARI is always on? For ospray, I think it can be enabled and disabled at run time too.

griffin28 replied:
Good catch @markcmiller86. This should be updated to test for ANARI being enabled at runtime. You have to explicitly enable ANARI to change the rendering pass so ANARI is not always on. It can be turned off **AND** on at runtime. I'll need to add back `useAnari` that gets set to the same value as `RenderingAttributes::anariRendering`.

markcmiller86 commented on src/avt/VisWindow/VisWindow/VisWindow.C:
The diffs GitHub was showing me for this file were confusing me. It was showing for example `::FullFrameOff` was replaced with `::SetAnariSPP`. But, when I look at the full file, it looks fine :confused:.

markcmiller86 commented on src/bin/internallauncher:
I may be confused again but this file looks like it includes changes made by @biagas. Is this perhaps related to rebasing?

markcmiller86 commented on src/common/state/RenderingAttributes.C:
This is an attributes object. So, I am not sure we support conditionally compiled members here but there are a ton of things related to Anari here that, in theory anyways, should only be there if anari is being used.

I wonder if it makes sense to make an AttributeGroup submember of this object that holds all the anari stuff and that whole subobject is conditionally compil

markcmiller86 commented on src/common/utility/StringHelpers.C:
Just noting another case of changes from another developer. I think this has to do with rebasing.

markcmiller86 reviewed (commented):
@griffin28 I am about 1/2 way through and will push the **Submit review** button now with a handful of remarks so far. Will finish up next week.

markcmiller86 commented on src/tools/dev/diagnostics/CMakeLists.txt:
```diff
@@ -15,7 +15,8 @@ add_subdirectory(networktest)
 
 
-if(HAVE_OSMESA AND VTK_VERSION VERSION_LESS "9.4.0")
+# OSMesa should always be available with VTK 9.4.1
```
The previous code indicated "9.4.0". Just wanna confirm...was that just an error and the correct version is "9.4.1"

biagas replied:
@markcmiller86 We have never released a version of VisIt using 9.4.x yet, so I think specifying the current version we are using is the correct way to go.  This was part of a PR I did awhile ago, and not part of @griffin28's Anari work.

markcmiller86 reviewed (commented):
Ok @griffin28 I've finished all the files.

@biagas I am wondering about a couple of cases here where the Anari controls, which number around 40, are put at the same level as other, similar controls (RenderingAttributes, VisWindow). On the one hand, this makes perfect sense semantically. On the other hand, I wonder if the right thing to do is create a sub-object of some kind for these to better isolate them to their own pieces.

biagas commented:
> @biagas I am wondering about a couple of cases here where the Anari controls, which number around 40, are put at the same level as other, similar controls (RenderingAttributes, VisWindow). On the one hand, this makes perfect sense semantically. On the other hand, I wonder if the right thing to do is create a sub-object of some kind for these to better isolate them to their own pieces.

I am not sure what you are questioning/suggesting.

markcmiller86 commented:
> I am not sure what you are questioning/suggesting.

Repeating the last sentence in my remark... I wonder if the right thing to do is create a sub-object of some kind for *these slew of Anari controls* cases to better isolate them to their own pieces.

griffin28 commented:
> 
> Repeating the last sentence in my remark... I wonder if the right thing to do is create a sub-object of some kind for _these slew of Anari controls_ cases to better isolate them to their own pieces.

@markcmiller86 this was the motivation for the dynamic UI solution. Most of those will go away and be collapsed into:
```
stringVector   anariRendererParameters;
stringVector   anariUSDParameters;
```
The **OSPRay** attributes are a similar case.

markcmiller86 commented:
@griffin28 I hope we haven't delayed you here. Its fallen of my radar. Let us know when you want another look

griffin28 commented:
@markcmiller86 no delay at all. The runtime testing of ANARI turned out to be a little trickier than expected. I'm going to submit a new PR with the current changes minus all of the commits from other developers. This will happen tomorrow or later this week.

griffin28 committed:
Added runtime check for ANARI

griffin28 commented:
Creating a new PR to avoid potential conflicts from the re-base.
